Remove GIL from RRefContext

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32807 Remove GIL from RRefContext**

After this commit, RRefContext no longer depends on pybind.

Differential Revision: [D19636316](https://our.internmc.facebook.com/intern/diff/D19636316)

mrshenli committed:
Remove GIL from RRefContext

After this commit, RRefContext no longer depends on pybind.

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 02ad60ea:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=02ad60eaed597906777d783fc6a89405c04dfbe0).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4377463) pytorch_linux_backward_compatibility_check_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=93003018) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=93003018))
<details>
<summary>
<code>Jan 30 03:08:47 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.</code>
</summary>

```
Jan 30 03:08:47 processing existing schema:  aten::values(Dict(str, t) self) -> (t[](*)) 
Jan 30 03:08:47 processing existing schema:  aten::values(Dict(int, t) self) -> (t[](*)) 
Jan 30 03:08:47 processing existing schema:  aten::values(Dict(float, t) self) -> (t[](*)) 
Jan 30 03:08:47 processing existing schema:  aten::values(Dict(Tensor, t) self) -> (t[](*)) 
Jan 30 03:08:47 processing existing schema:  aten::view(Tensor(a) self, int[] size) -> (Tensor(a)) 
Jan 30 03:08:47 processing existing schema:  aten::where(Tensor condition) -> (Tensor[]) 
Jan 30 03:08:47 processing existing schema:  aten::where.self(Tensor condition, Tensor self, Tensor other) -> (Tensor) 
Jan 30 03:08:47 processing existing schema:  aten::zeros_like(Tensor self, *, int? memory_format=None) -> (Tensor) 
Jan 30 03:08:47 processing existing schema:  aten::zeros_like.dtype(Tensor self, *, int dtype, int layout, Device device, bool pin_memory=False, int? memory_format=None) -> (Tensor) 
Jan 30 03:08:47 processing existing schema:  prim::Constant() -> (...) 
Jan 30 03:08:47 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.  
Jan 30 03:08:47  
Jan 30 03:08:47 Broken ops: [ 
Jan 30 03:08:47 	quantized::linear_relu_dynamic_fp16(Tensor X, Tensor W_prepack) -> (Tensor Y) 
Jan 30 03:08:47 ] 
Jan 30 03:08:47 + cleanup 
Jan 30 03:08:47 + retcode=1 
Jan 30 03:08:47 + set +x 
Jan 30 03:08:47 =================== sccache compilation log =================== 
Jan 30 03:08:47 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 30 03:08:47 Compile requests                 0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

xush6528 commented:
Looks good. Can you apply the internal formatter/linter?

xush6528 commented:
Can you remove "#include <torch/csrc/utils/pybind.h>" in "rref_context.h" as well

zhaojuanmao commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -269,7 +269,12 @@ If the future completes with an error, an exception is thrown.
+    auto deletedRRefs =
+        RRefContext::getInstance().destroyInstance(ignoreRRefLeak);
+    if (!deletedRRefs.empty()) {
+      pybind11::gil_scoped_acquire ag;
```
nit: no need GIL here? as pybind implicitly has GIL?

zhaojuanmao reviewed (approved):
Thanks for removing this python dependency! see one small nit

Maybe we can move rref_context.cpp from libtorch_python_sources to libtorch_sources in build_variables.py and CMakeLists.txt?

mrshenli commented:
> Can you remove "#include <torch/csrc/utils/pybind.h>" in "rref_context.h" as well?

the pybind header is actually imported in `rref_impl.h`, which can be removed after #32753

mrshenli commented:
> Maybe we can move rref_context.cpp from libtorch_python_sources to libtorch_sources in build_variables.py and CMakeLists.txt?

This needs to come in a separate PR after #32753, as rref_context.h depends on rref_impl.h. We need to move both of them to libtorch_sources together.

mrshenli committed:
Update on "Remove GIL from RRefContext"


After this commit, RRefContext no longer depends on pybind.

Differential Revision: [D19636316](https://our.internmc.facebook.com/intern/diff/D19636316)

[ghstack-poisoned]
